### PR TITLE
Add --header/-H to dapr invoke

### PR DESCRIPTION
**Pull request by hunter007**
# Description

to support pass customized headers to dapr app

See https://github.com/dapr/go-sdk/pull/405

## Issue reference

Fix https://github.com/dapr/cli/issues/1286

### CODE REVIEW

**Review by mukundansundar — COMMENTED** [review]
@hunter007 Thanks for adding this feature.
Can you add E2E for the this feature please?

**Comment by hunter007** [comment]
> @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?

:)  Review this PR first: https://github.com/dapr/go-sdk/pull/405

**Comment by mukundansundar** [comment]
> > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> 
> :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)

seems the PR has been merged :)

**Comment by hunter007** [comment]
> > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > 
> > 
> > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> 
> seems the PR has been merged :)



> > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > 
> > 
> > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> 
> seems the PR has been merged :)

Yes,  so the PR could be reviewd.

**Comment by mukundansundar** [comment]
> > > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > > 
> > > 
> > > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> > 
> > 
> > seems the PR has been merged :)
> 
> > > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > > 
> > > 
> > > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> > 
> > 
> > seems the PR has been merged :)
> 
> Yes, so the PR could be reviewd.

Won't the dependency need to be updated to point to the latest go-sdk in this case?

**Comment by hunter007** [comment]
> > > > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > > > 
> > > > 
> > > > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> > > 
> > > 
> > > seems the PR has been merged :)
> > 
> > 
> > > > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > > > 
> > > > 
> > > > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> > > 
> > > 
> > > seems the PR has been merged :)
> > 
> > 
> > Yes, so the PR could be reviewd.
> 
> Won't the dependency need to be updated to point to the latest go-sdk in this case?

no need.

> > > > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > > > 
> > > > 
> > > > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> > > 
> > > 
> > > seems the PR has been merged :)
> > 
> > 
> > > > > @hunter007 Thanks for adding this feature. Can you add E2E for the this feature please?
> > > > 
> > > > 
> > > > :) Review this PR first: [dapr/go-sdk#405](https://github.com/dapr/go-sdk/pull/405)
> > > 
> > > 
> > > seems the PR has been merged :)
> > 
> > 
> > Yes, so the PR could be reviewd.
> 
> Won't the dependency need to be updated to point to the latest go-sdk in this case?

No need. https://github.com/dapr/go-sdk/pull/405 is used when writing app with go-sdk. Without https://github.com/dapr/go-sdk/pull/405, `cli invoke` will still send HTTP header by `-H` to dapr app

**Comment by mukundansundar** [comment]
@hunter007 the specific test that you have written seems to fail `output does not contain "aValue"` ... Can you take a look into this?

**Comment by hunter007** [comment]
> @hunter007 the specific test that you have written seems to fail `output does not contain "aValue"` ... Can you take a look into this?

It dependent on dapr/go-sdk which's version is after v1.7.0.

**Comment by mukundansundar** [comment]
> > @hunter007 the specific test that you have written seems to fail `output does not contain "aValue"` ... Can you take a look into this?
> 
> It dependent on dapr/go-sdk which's version is after v1.7.0.

This PR needs to be modified to use the latest go-sdk then in tests.

**Comment by mukundansundar** [comment]
@hunter007 1.8 version of go-sdk was released and I believe that needs to be used for the header functionality to work.

**Comment by hunter007** [comment]
@mukundansundar github action timeouted, have a retry?

File: cmd/invoke.go [hunk 1]
```diff
@@ -98,6 +121,7 @@ func init() {
 	InvokeCmd.Flags().StringVarP(&invokeData, "data", "d", "", "The JSON serialized data string (optional)")
 	InvokeCmd.Flags().StringVarP(&invokeVerb, "verb", "v", defaultHTTPVerb, "The HTTP verb to use")
 	InvokeCmd.Flags().StringVarP(&invokeDataFile, "data-file", "f", "", "A file containing the JSON serialized data (optional)")
+	InvokeCmd.Flags().StringArrayVarP(&invokeHeaders, "header", "H", []string{}, "0 or more HTTP Header")
```
Comment (mukundansundar) [inline cmd/invoke.go:124]:
```suggestion
	InvokeCmd.Flags().StringArrayVarP(&invokeHeaders, "header", "H", []string{}, "HTTP headers to be used on invoke")
```

File: cmd/invoke.go [hunk 2]
```diff
@@ -43,6 +45,9 @@ var InvokeCmd = &cobra.Command{
 # Invoke a sample method on target app with POST Verb
 dapr invoke --app-id target --method sample --data '{"key":"value"}
 
+# Invoke a sample method on target app with customized Header
```
Comment (mukundansundar) [inline cmd/invoke.go:48]: for multiple headers, is it going to be comma separated values?

File: pkg/standalone/invoke_test.go [hunk 3]
```diff
@@ -69,8 +74,23 @@ func TestInvoke(t *testing.T) {
 			},
 			expectedPath: "/v1.0/invoke/testapp/method/test",
 			postResponse: "test payload",
+			header:       nil,
 			resp:         "successful invoke",
 		},
+		{
+			name:   "appID found successful invoke with customized header",
```
Comment (mukundansundar) [inline pkg/standalone/invoke_test.go:81]: please add one test with multiple header values.

**Review by mukundansundar — CHANGES_REQUESTED** [review]
small changes please.

**Pull request closed without merging**